README.md

```diff
@@ -29,20 +29,67 @@ describe('Blog page', () => {
 
 ## What
 
-The idea behind this library is to enable DOM integration tests on Next.js pages along with [data fetching][next-docs-data-fetching] and [routing][next-docs-routing].
+The idea behind this library is to reproduce as closely as possible the way Next.js works without spinning up servers, and render the output in a local JSDOM environment.
 
-The testing approach suggested here consists of manually mocking external API's dependencies and get the component instance matching a given route.
+In order to provide a valuable testing experience `next-page-tester` replicates the **rendering flow of an actual `next.js` application**:
 
-Next page tester will take care of:
+1. **fetch data** for a given route
+2. **render** the server-side rendered result to JSDOM as plain html (including `head` element)
+3. **mount/hydrate** the client application to the previously rendered html
 
+The mounted application is **interactive** and can be tested with any DOM testing library (like [`@testing-library/dom`](https://testing-library.com/docs/dom-testing-library/intro)).
+
+`next-page-tester` will take care of:
+
+- load and execute modules in the **expected browser or server environments**
 - **resolving** provided **routes** into matching page component
 - calling **Next.js data fetching methods** (`getServerSideProps`, `getInitialProps` or `getStaticProps`) if the case
 - set up a **mocked `next/router` provider** initialized with the expected values (to test `useRouter` and `withRouter`)
 - wrapping page with custom `_app` and `_document` components
-- **instantiating** page component with **expected page props**
-- emulating client side navigation via `Link`, `router.push`, `router.replace`
+- emulating **client side navigation** via `Link`, `router.push`, `router.replace`
 - handling pages' `redirect` returns
 
+## API
+
+### getPage
+
+`getPage` accepts an [option object](#options) and returns 4 values:
+
+```js
+import { getPage } from 'next-page-tester';
+const { render, renderHTML, html, page } = await getPage({ options });
+```
+
+### render()
+
+Type: `() => HTMLElement<NextRoot>`<br/>
+Returns: `#__next` root element element.
+
+You will mostly use this render method. It calls `renderHTML()` and then mounts/hydrates the React application into JSDOM `#__next` root element. This is what the user gets when she visits a page.
+
+### renderHTML()
```

Comment:
What do you think about this?

```suggestion
### prepareDocument()
```

Reply:
I'll try to put down what I think `renderHTML` does compared to a real Next.js app.

From my point of `renderHTML()/whatever` represents the usual initial rendering process which happens in any browser when a user visits any HTML page: HTML file download, parsed and a DOM tree generated out if it.

For a static HTML page the game ends with `renderHTML()/whatever`.

The exact same thing happens for a SSRendered React app but there's is an extra step: when JS resources are loaded React takes over, and does whatever it was told to do with the just rendered DOM.

From users' perspective, I'd say that `renderHTML()` renders the `html` string in the DOM (or any equivalent word for transforming a string into it's visual representation)

`prepareDocument` sounds a bit like "setting up the environment before we do anything", while `renderHTML()` paints the whole page markup on the screen (except for the detail that JSDOM doesn't provide visual output 😄 ).

`renderHTML()` should be something close to:
- `renderSSRResultInDOM`
- `renderSSR`
- `renderStaticHTML`
- `renderSSRHTML`

`render()` should be something like:
- `renderAndMountPage`
- `renderPageAndMountReact`
- `renderSSRAndMountPage`
- `renderAndHydrate`

Reply:
`renderStaticMarkup` & `render`?

I think `renderStaticMarkup` makes it clear that output is SSR and that it's not interactive. As for the "main" `render` method I would not complicate and make name as simple as possible: `render`.

Reply:
I'd slim down the list to:
- renderStaticMarkup
- renderStaticDocument
- renderDocument

Taken from React docs:
> Render a React element to its initial HTML. React will return an HTML string. You can use this method to generate HTML on the server and send the markup down on the initial request for faster page loads and to allow search engines to crawl your pages for SEO purposes.

> If you call ReactDOM.hydrate() on a node that already has this server-rendered markup, React will preserve it and only attach event handlers, allowing you to have a very performant first-load experience.

- renderInitialHTML
- renderInitialMarkup
- renderServerRenderedMarkup
- renderServerRenderedHTML

(even though our server-rendered is emulated)

I'd try not to add another return method, right now. I don't see much value right now. Actually I considered removing both `html` and `page`: they have sense only if someone tried to render and hydrate on their side.